extension/ts/main/adapters/tags/parseTags.ts

```diff
@@ -0,0 +1,32 @@
+import {TagNode, TagTree} from "./types";
+
+interface ParserOptions {
+	rows: string[][];
+	fromRow: number;
+	depth: number;
+	tree: TagTree;
+	parent?: TagNode;
+}
+
+const parseRows = ({rows, fromRow, depth, tree, parent}: ParserOptions): number => {
+	let row = fromRow;
+	while (row < rows.length) {
+		const tag = rows[row][depth];
+		if (tag) {
+			const node = {tag, parent};
+			tree.set(tag, node);
+			row = parseRows({rows, fromRow: row + 1, depth: depth + 1, tree, parent: node});
+		} else {
+			break;
+		}
+	}
+	return row;
+};
+
+const parseTree = (rows: string[][]) => {
+	const tree: TagTree = new Map();
+	for (let fromRow = 0; fromRow < rows.length; fromRow = parseRows({rows, fromRow, depth: 0, tree}) + 1);
```

Comment:
weird ass for loop but i respect it

Reply:
should have been recursive but now i don't wanna touch any of this lol